waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,605 @@
+                             (for [i (range ports)]
+                               {:name (str "PORT" i) :value (str (+ port0 i))})))
+        params {:k8s-name (service-id->k8s-name scheduler service-id)
+                :backend-protocol backend-proto
+                :backend-protocol-caps (string/upper-case backend-proto)
+                :cmd cmd
+                :cpus cpus
+                :env template-env
+                :grace-period-secs grace-period-secs
+                :health-check-interval-secs health-check-interval-secs
+                :health-check-max-consecutive-failures health-check-max-consecutive-failures
+                :health-check-url (sd/service-description->health-check-url service-description)
+                :home-path home-path
+                :memory  (str mem "Mi")
+                :min-instances min-instances
+                :port-count ports
+                :run-as-user run-as-user
+                :service-id service-id
+                :ssl? (= "https" backend-proto)}
+        edn-opts {:readers {'waiter/param params
+                            'waiter/param-str (comp str params)
+                            'waiter/port #(+ port0 %)
+                            'waiter.fn/into (fn [[xs ys]] (into xs ys))
+                            'waiter.fn/lower-case string/upper-case
+                            'waiter.fn/str #(apply str %)
+                            'waiter.fn/upper-case string/upper-case}}]
+    (try
+      (->> rs-spec-file-path slurp (edn/read-string edn-opts))
+      (catch Throwable e
+        (log/error e "Error creating ReplicaSet specification for" service-id)
+        (throw e)))))
+
+(defn- create-service
+  "Reify a Waiter Service as a Kuberentes ReplicaSet."
+  [service-id descriptor {:keys [api-server-url http-client] :as scheduler} service-id->password-fn]
+  (let [{:strs [run-as-user] :as service-description} (:service-description descriptor)
+        spec-json (service-spec scheduler service-id service-description service-id->password-fn)
+        request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (service-description->namespace service-description)
+                         "/replicasets")
+        response-json (api-request http-client request-url
+                                   :body (as-json spec-json)
+                                   :request-method :post)]
+    (when-not (= "ReplicaSet" (:kind response-json))
+      (log/error "Invalid response from ReplicaSet create request"
+                 (as-json response-json)))
+    (replicaset->Service response-json)))
+
+(defn- delete-service
+  "Delete the Kubernetes ReplicaSet corresponding to a Waiter Service.
+   Ensures that all controlled Pods are deleted before removing the ReplicaSet.
+   This operation will leave Waiter in a sane state on a failure."
+  [{:keys [api-server-url http-client] :as scheduler} service]
+  (when-not service
```

Comment:
Why can this happen?

Reply:
The `service-id->service` call that feeds into this can return `nil` if it fails (e.g., 404).

Reply:
Should the delete call be avoided when that happens?

Reply:
Done. Now `service-id->service` throws on 404 errors.